beacon/light/request/scheduler.go

```diff
@@ -0,0 +1,183 @@
+	// registered in.
+	Process(ServerSet, []ServerEvent) bool
+}
+
+// Scheduler is a modular network data retrieval framework that coordinates multiple
+// servers and retrieval mechanisms (modules). It implements a trigger mechanism
+// that calls the Process function of registered modules whenever either the state
+// of existing data structures or connected servers could allow new operations.
+type Scheduler struct {
+	lock    sync.Mutex
+	clock   mclock.Clock
+	modules []Module // first has highest priority
+	servers map[Server]struct{}
+	stopCh  chan chan struct{}
+
+	triggerCh chan struct{} // restarts waiting sync loop
+	//	testWaitCh       chan struct{} // accepts sends when sync loop is waiting
+	//	testTimerResults []bool        // true is appended when simulated timer is processed; false when stopped
+	events []ServerEvent
+}
+
+type ServerEvent struct {
+	Server Server
+	Event  Event
+}
+
+// NewScheduler creates a new Scheduler.
+func NewScheduler(clock mclock.Clock) *Scheduler {
+	s := &Scheduler{
+		clock:   clock,
+		servers: make(map[Server]struct{}),
+		stopCh:  make(chan chan struct{}),
+		// Note: testWaitCh should not have capacity in order to ensure
+		// that after a trigger happens testWaitCh will block until the resulting
+		// processing round has been finished
+		triggerCh: make(chan struct{}, 1),
+		//testWaitCh: make(chan struct{}),
+	}
+	return s
+}
+
+// RegisterModule registers a module. Should be called before starting the scheduler.
+// In each processing round the order of module processing depends on the order of
+// registration.
+func (s *Scheduler) RegisterModule(m Module) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.modules = append(s.modules, m)
+}
+
+// RegisterServer registers a new server.
+func (s *Scheduler) RegisterServer(server Server) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	server.Subscribe(func(event Event) {
+		s.handleEvent(server, event)
+	})
+	s.servers[server] = struct{}{}
+	s.Trigger()
+}
+
+// UnregisterServer removes a registered server.
+func (s *Scheduler) UnregisterServer(server Server) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	server.Unsubscribe()
+	delete(s.servers, server)
+}
+
+// Start starts the scheduler. It should be called after registering all modules
+// and before registering any servers.
+func (s *Scheduler) Start() {
+	go s.syncLoop()
+}
+
+// Stop stops the scheduler.
+func (s *Scheduler) Stop() {
+	s.lock.Lock()
+	for server, _ := range s.servers {
+		server.Unsubscribe()
+	}
+	s.servers = nil
+	s.lock.Unlock()
+	stop := make(chan struct{})
+	s.stopCh <- stop
+	<-stop
+}
+
+// syncLoop calls all processable modules in the order of their registration.
+// A round of processing starts whenever there is at least one processable module.
+// Triggers triggered during a processing round do not affect the current round
+// but ensure that there is going to be a next round.
+func (s *Scheduler) syncLoop() {
+	for {
+		s.lock.Lock()
+		s.processModules()
```

Comment:
the processModules is holding `s.lock`. If any of these generic modules try to interact with the scheduler, e..g registering a new server, this will deadlock, no?